src/index.js

```diff
@@ -1,169 +1,323 @@
+	 * @returns
+	 */
+	maybeScrollToAnchor(hash, context) {
+		// Bail early if the hash is null
+		if (hash == null) {
+			return false;
+		}
+		const element = this.getAnchorElement(hash);
+		if (!element) {
+			console.warn(`Element ${hash} not found`);
+			return false;
+		}
+		if (!(element instanceof Element)) {
+			console.warn(`Element ${hash} is not a DOM node`);
+			return false;
+		}
+		const top =
+			element.getBoundingClientRect().top + window.pageYOffset - this.getOffset(element);
+		this.swup.scrollTo(top, this.shouldAnimate(context));
+		return true;
+	}
+
+	/**
+	 * Handles `transitionStart`
+	 * @param {PopStateEvent} popstate
+	 */
+	onTransitionStart = (popstate) => {
+		if (this.options.doScrollingRightAway && !this.swup.scrollToElement) {
+			this.doScrollingBetweenPages(popstate);
+		}
+	};
+
+	/**
+	 * Handles `contentReplaced`
+	 * @param {PopStateEvent} popstate
+	 */
+	onContentReplaced = (popstate) => {
+		if (!this.options.doScrollingRightAway || this.swup.scrollToElement) {
+			this.doScrollingBetweenPages(popstate);
+		}
+
+		this.restoreScrollContainers(popstate);
+	};
+
+	/**
+	 * Scrolls the window, based on context
+	 *
+	 * @param {PopStateEvent|boolean} popstate
+	 * @returns void
+	 */
+	doScrollingBetweenPages = (popstate) => {
+		const swup = this.swup;
+
+		// Bail early on popstate and inactive `animateHistoryBrowsing`
+		if (popstate && !swup.options.animateHistoryBrowsing) {
+			return;
+		}
+
+		// Try scrolling to a given anchor
+		if (this.maybeScrollToAnchor(swup.scrollToElement, 'betweenPages')) {
+			swup.scrollToElement = null;
+			return;
+		}
+
+		// Finally, scroll to either the stored scroll position or to the very top of the page
+		const scrollPositions = this.getStoredScrollPositions(getCurrentUrl());
+		const top = (scrollPositions.window && scrollPositions.window.top) || 0;
+		swup.scrollTo(top, this.shouldAnimate('betweenPages'));
+	};
+
+	/**
+	 * Stores the current scroll positions for the URL we just came from
+	 */
+	onWillReplaceContent = () => {
+		this.storeScrollPositions(this.previousUrl);
+		this.previousUrl = getCurrentUrl();
+	};
+
+	/**
+	 * Deletes the scroll positions for the URL a link is pointing to,
+	 * if shouldRestoreScrollPosition doesn't evaluate to true
+	 *
+	 * @param {event} e
+	 * @returns void
+	 */
+	onClickLink = (e) => {
+		if (!this.options.shouldRestoreScrollPosition(e.delegateTarget)) {
+			return;
+		}
+		const url = new Link(e.delegateTarget).getAddress();
+		this.deleteStoredScrollPositions(url);
+	};
+
+	/**
+	 * Stores the scroll positions for the current URL
+	 * @param {string} url
+	 */
+	storeScrollPositions(url) {
+		// create an object to store the scroll positions
+		const storeEntry = {
+			window: { top: window.scrollY, left: window.scrollX },
+			containers: []
+		};
+		// fill up the object with scroll positions for each matching element
+		const $containers = document.querySelectorAll(this.options.scrollContainers);
```

Comment:
I don't think the codebase uses the `$` prefix for dom elements, I would keep that consistent.